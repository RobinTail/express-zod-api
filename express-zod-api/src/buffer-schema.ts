import { z } from "zod/v4";

export const bufferSchema = z.custom<Buffer>(
  (subject) => Buffer.isBuffer(subject),
  { error: "Expected Buffer" },
);
