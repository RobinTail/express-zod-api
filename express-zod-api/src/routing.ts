import { IRouter, RequestHandler } from "express";
import createHttpError from "http-errors";
import { isProduction } from "./common-helpers";
import { CommonConfig } from "./config-type";
import { ContentType } from "./content-type";
import { Diagnostics } from "./diagnostics";
import { AuxMethod, Method } from "./method";
import { Routable } from "./routable";
import { OnEndpoint, walkRouting } from "./routing-walker";
import { GetLogger } from "./server-helpers";

export interface Routing {
  [SEGMENT: string]: Routing | Routable;
}

export type Parsers = Partial<Record<ContentType, RequestHandler[]>>;

/** @link https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/405 */
export const createWrongMethodHandler =
  (allowedMethods: Array<Method | AuxMethod>): RequestHandler =>
  ({ method }, res, next) => {
    const Allow = allowedMethods.join(", ").toUpperCase();
    res.set({ Allow }); // in case of a custom errorHandler configured that does not care about headers in error
    const error = createHttpError(405, `${method} is not allowed`, {
      headers: { Allow },
    });
    next(error);
  };

export const initRouting = ({
  app,
  getLogger,
  config,
  routing,
  parsers,
}: {
  app: IRouter;
  getLogger: GetLogger;
  config: CommonConfig;
  routing: Routing;
  parsers?: Parsers;
}) => {
  let doc: Diagnostics | undefined = new Diagnostics(getLogger()); // disposable instance
  const familiar = new Map<string, Array<Method | AuxMethod>>();
  const onEndpoint: OnEndpoint = (endpoint, path, method, siblingMethods) => {
    if (!isProduction()) {
      doc?.checkJsonCompat(endpoint, { path, method });
      doc?.checkPathParams(path, endpoint, { method });
    }
    const matchingParsers = parsers?.[endpoint.getRequestType()] || [];
    const handler: RequestHandler = async (request, response) => {
      const logger = getLogger(request);
      if (config.cors) {
        const accessMethods: Array<Method | AuxMethod> = [
          method,
          ...(siblingMethods || []),
          "options",
        ];
        const methodsLine = accessMethods.join(", ").toUpperCase();
        const defaultHeaders: Record<string, string> = {
          "Access-Control-Allow-Origin": "*",
          "Access-Control-Allow-Methods": methodsLine,
          "Access-Control-Allow-Headers": "content-type",
        };
        const headers =
          typeof config.cors === "function"
            ? await config.cors({ request, endpoint, logger, defaultHeaders })
            : defaultHeaders;
        for (const key in headers) response.set(key, headers[key]);
      }
      return endpoint.execute({ request, response, logger, config });
    };
    if (!familiar.has(path)) {
      familiar.set(path, []);
      if (config.cors) {
        app.options(path, ...matchingParsers, handler);
        familiar.get(path)?.push("options");
      }
    }
    familiar.get(path)?.push(method);
    app[method](path, ...matchingParsers, handler);
  };
  walkRouting({ routing, onEndpoint, onStatic: app.use.bind(app) });
  doc = undefined; // hint for garbage collector
  if (config.wrongMethodBehavior !== 405) return;
  for (const [path, allowedMethods] of familiar.entries())
    app.all(path, createWrongMethodHandler(allowedMethods));
};
