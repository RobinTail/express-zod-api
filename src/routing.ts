import { IRouter } from "express";
import { CommonConfig } from "./config-type";
import { DependsOnMethod } from "./depends-on-method";
import { AbstractEndpoint } from "./endpoint";
import { AbstractLogger } from "./logger";
import { walkRouting } from "./routing-walker";
import { ServeStatic } from "./serve-static";
import { getStartupLogo } from "./startup-logo";

export interface Routing {
  [SEGMENT: string]: Routing | DependsOnMethod | AbstractEndpoint | ServeStatic;
}

export const initRouting = ({
  app,
  logger,
  config,
  routing,
}: {
  app: IRouter;
  logger: AbstractLogger;
  config: CommonConfig;
  routing: Routing;
}) => {
  if (config.startupLogo !== false) {
    console.log(getStartupLogo());
  }
  logger.debug("Running", process.env.TSUP_BUILD || "from sources");
  walkRouting({
    routing,
    hasCors: !!config.cors,
    onEndpoint: (endpoint, path, method) => {
      app[method](path, async (request, response) => {
        logger.info(`${request.method}: ${path}`);
        await endpoint.execute({ request, response, logger, config });
      });
    },
    onStatic: (path, handler) => {
      app.use(path, handler);
    },
  });
};
