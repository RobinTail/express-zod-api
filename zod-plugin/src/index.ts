import "./runtime.ts"; // side effects here
export { pack, unpack } from "./packer.ts";
export { getBrand } from "./brand.ts";
