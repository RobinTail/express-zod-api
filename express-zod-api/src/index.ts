import "@express-zod-api/zod-plugin"; // side effects here
export { createConfig } from "./config-type.ts";
export {
  EndpointsFactory,
  defaultEndpointsFactory,
  arrayEndpointsFactory,
} from "./endpoints-factory.ts";
export { getMessageFromError } from "./common-helpers.ts";
export { ensureHttpError } from "./result-helpers.ts";
export { BuiltinLogger } from "./builtin-logger.ts";
export { Middleware } from "./middleware.ts";
export {
  ResultHandler,
  defaultResultHandler,
  arrayResultHandler,
} from "./result-handler.ts";
export { DependsOnMethod } from "./depends-on-method.ts";
export { ServeStatic } from "./serve-static.ts";
export { createServer, attachRouting } from "./server.ts";
export { Documentation } from "./documentation.ts";
export {
  DocumentationError,
  RoutingError,
  OutputValidationError,
  InputValidationError,
  MissingPeerError,
} from "./errors.ts";
export { testEndpoint, testMiddleware } from "./testing.ts";
export { Integration } from "./integration.ts";
export { EventStreamFactory } from "./sse.ts";

export { ez } from "./proprietary-schemas.ts";

// Convenience types
export type { Depicter } from "./documentation-helpers.ts";
export type { Producer } from "./zts-helpers.ts";

// Interfaces exposed for augmentation
export type { LoggerOverrides } from "./logger-helpers.ts";
export type { TagOverrides } from "./common-helpers.ts";

// Issues 952, 1182, 1269: Insufficient exports for consumer's declaration
import type {} from "qs"; // fixes TS2742 for attachRouting
export type { Routing } from "./routing.ts";
export type { FlatObject } from "./common-helpers.ts";
export type { Method } from "./method.ts";
export type { IOSchema } from "./io-schema.ts";
export type { CommonConfig, AppConfig, ServerConfig } from "./config-type.ts";
export type { ApiResponse } from "./api-response.ts";
export type {
  BasicSecurity,
  BearerSecurity,
  CookieSecurity,
  HeaderSecurity,
  InputSecurity,
  OAuth2Security,
  OpenIdSecurity,
} from "./security.ts";
