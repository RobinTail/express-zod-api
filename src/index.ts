import "./zod-plugin";

export { createConfig } from "./config-type";
export { AbstractEndpoint } from "./endpoint";
export {
  EndpointsFactory,
  defaultEndpointsFactory,
  arrayEndpointsFactory,
} from "./endpoints-factory";
export {
  getExamples,
  getMessageFromError,
  getStatusCodeFromError,
} from "./common-helpers";
export { createLogger } from "./logger";
export { createMiddleware } from "./middleware";
export {
  createResultHandler,
  defaultResultHandler,
  arrayResultHandler,
} from "./result-handler";
export { DependsOnMethod } from "./depends-on-method";
export { ServeStatic } from "./serve-static";
export { createServer, attachRouting } from "./server";
export { Documentation } from "./documentation";
export {
  DocumentationError,
  RoutingError,
  OutputValidationError,
  InputValidationError,
  MissingPeerError,
} from "./errors";
export { testEndpoint } from "./testing";
export { Integration } from "./integration";

export { ez } from "./proprietary-schemas";

// Convenience types
export type { Depicter } from "./documentation-helpers";
export type { Producer } from "./zts-helpers";

// Issues 952, 1182, 1269: Insufficient exports for consumer's declaration
export type { MockOverrides } from "./testing";
export type { Routing } from "./routing";
export type { LoggerOverrides } from "./logger";
export type { FlatObject } from "./common-helpers";
export type { Method } from "./method";
export type { IOSchema } from "./io-schema";
export type { CommonConfig, AppConfig, ServerConfig } from "./config-type";
export type { MiddlewareDefinition } from "./middleware";
export type { ResultHandlerDefinition } from "./result-handler";
export type { ApiResponse } from "./api-response";
export type {
  BasicSecurity,
  BearerSecurity,
  CookieSecurity,
  CustomHeaderSecurity,
  InputSecurity,
  OAuth2Security,
  OpenIdSecurity,
} from "./security";
