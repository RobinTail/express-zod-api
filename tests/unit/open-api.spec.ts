import {routing} from '../../example/routing';
import {z, OpenAPI, defaultEndpointsFactory} from '../../src';
import {expectType} from 'tsd';

describe('Open API generator', () => {
  describe('generateOpenApi()', () => {
    test('should generate the correct schema of example routing', () => {
      const spec = new OpenAPI({
        routing,
        version: '1.2.3',
        title: 'Example API',
        serverUrl: 'http://example.com'
      }).getSpecAsYaml();
      expect(spec).toMatchSnapshot();
    });

    test('should generate the correct schema for complex types', () => {
      const literalValue = 'something' as const;
      const spec = new OpenAPI({
        routing: {
          v1: {
            getSomething: defaultEndpointsFactory.build({
              methods: ['get'],
              input: z.object({
                array: z.array(z.number().int().positive()).min(1).max(3),
                unlimited: z.array(z.boolean()),
                transformer: z.string().transform((str) => str.length)
              }),
              output: z.object({
                literal: z.literal(literalValue),
                transformation: z.number(),
              }),
              handler: async ({input}) => ({
                literal: literalValue,
                transformation: input.transformer,
              })
            })
          }
        },
        version: '3.4.5',
        title: 'Testing Complex Types',
        serverUrl: 'http://example.com'
      }).getSpecAsYaml();
      expect(spec).toMatchSnapshot();
    });

    test('should generate the correct schema for nullable and optional types', () => {
      const spec = new OpenAPI({
        routing: {
          v1: {
            getSomething: defaultEndpointsFactory.build({
              methods: ['get'],
              input: z.object({
                optional: z.string().optional(),
                optDefault: z.string().optional().default('test'),
                nullish: z.boolean().nullish(),
                nuDefault: z.number().int().positive().nullish().default(123)
              }),
              output: z.object({
                nullable: z.string().nullable(),
              }),
              handler: async () => ({
                nullable: null,
              })
            })
          }
        },
        version: '3.4.5',
        title: 'Testing Nullable and Optional Types',
        serverUrl: 'http://example.com'
      }).getSpecAsYaml();
      expect(spec).toMatchSnapshot();
    });

    test('should generate the correct schema for intersection type', () => {
      const spec = new OpenAPI({
        routing: {
          v1: {
            getSomething: defaultEndpointsFactory.build({
              methods: ['post'],
              input: z.object({
                intersection: z.intersection(
                  z.object({
                    one: z.string(),
                  }),
                  z.object({
                    two: z.string(),
                  })
                ),
              }),
              output: z.object({
                and: z.object({
                  five: z.number().int().gte(0),
                }).and(z.object({
                  six: z.string()
                })),
              }),
              handler: async () => ({
                and: {
                  five: 5,
                  six: 'six'
                }
              })
            })
          }
        },
        version: '3.4.5',
        title: 'Testing Intersection and And types',
        serverUrl: 'http://example.com'
      }).getSpecAsYaml();
      expect(spec).toMatchSnapshot();
    });

    test('should generate the correct schema for union type', () => {
      const spec = new OpenAPI({
        routing: {
          v1: {
            getSomething: defaultEndpointsFactory.build({
              methods: ['post'],
              input: z.object({
                union: z.union([
                  z.object({
                    one: z.string(),
                    two: z.number().int().positive()
                  }),
                  z.object({
                    two: z.number().int().negative(),
                    three: z.string()
                  })
                ]),
              }),
              output: z.object({
                or: z.string().or(z.number().int().positive()),
              }),
              handler: async () => ({
                or: 554
              })
            })
          }
        },
        version: '3.4.5',
        title: 'Testing Union and Or Types',
        serverUrl: 'http://example.com'
      }).getSpecAsYaml();
      expect(spec).toMatchSnapshot();
    });

    test('should handle transformation schema in output', () => {
      const spec = new OpenAPI({
        routing: {
          v1: {
            getSomething: defaultEndpointsFactory.build({
              methods: ['post'],
              input: z.object({
                one: z.string(),
                two: z.number().int().positive()
              }),
              output: z.object({
                transform: z.string().transform((str) => str.length)
              }),
              handler: async () => ({
                transform: 'test'
              })
            })
          }
        },
        version: '3.4.5',
        title: 'Testing Transformation in response schema',
        serverUrl: 'http://example.com'
      }).getSpecAsYaml();
      expect(spec).toMatchSnapshot();
    });

    test('should handle bigint, boolean, date and null', () => {
      const spec = new OpenAPI({
        routing: {
          v1: {
            getSomething: defaultEndpointsFactory.build({
              method: 'post',
              input: z.object({
                bigint: z.bigint(),
                boolean: z.boolean(),
                date: z.date()
              }),
              output: z.object({
                null: z.null()
              }),
              handler: async () => ({
                null: null
              })
            })
          }
        },
        version: '3.4.5',
        title: 'Testing additional types',
        serverUrl: 'http://example.com'
      }).getSpecAsYaml();
      expect(spec).toMatchSnapshot();
    });

    test('should handle record', () => {
      const spec = new OpenAPI({
        routing: {
          v1: {
            getSomething: defaultEndpointsFactory.build({
              method: 'post',
              input: z.object({}),
              output: z.object({
                record: z.record(z.number().int()),
              }),
              handler: jest.fn()
            })
          }
        },
        version: '3.4.5',
        title: 'Testing record',
        serverUrl: 'http://example.com'
      }).getSpecAsYaml();
      expect(spec).toMatchSnapshot();
    });

    test('should handle type any', () => {
      const spec = new OpenAPI({
        routing: {
          v1: {
            getSomething: defaultEndpointsFactory.build({
              method: 'get',
              input: z.object({
                any: z.any()
              }),
              output: z.object({
                any: z.any()
              }),
              handler: jest.fn()
            })
          }
        },
        version: '3.4.5',
        title: 'Testing type any',
        serverUrl: 'http://example.com'
      }).getSpecAsYaml();
      expect(spec).toMatchSnapshot();
    });

    test('should handle different number types', () => {
      const spec = new OpenAPI({
        routing: {
          v1: {
            getSomething: defaultEndpointsFactory.build({
              method: 'post',
              input: z.object({
                double: z.number(),
                doublePositive: z.number().positive(),
                doubleNegative: z.number().negative(),
                doubleLimited: z.number().min(-0.5).max(0.5),
                int: z.number().int(),
                intPositive: z.number().int().positive(),
                intNegative: z.number().int().negative(),
                intLimited: z.number().int().min(-100).max(100),
                zero: z.number().int().nonnegative().nonpositive().optional(),
              }),
              output: z.object({
                bigint: z.bigint()
              }),
              handler: jest.fn()
            })
          }
        },
        version: '3.4.5',
        title: 'Testing numbers',
        serverUrl: 'http://example.com'
      }).getSpecAsYaml();
      expect(spec).toMatchSnapshot();
    });

    test('should handle different string types', () => {
      const spec = new OpenAPI({
        routing: {
          v1: {
            getSomething: defaultEndpointsFactory.build({
              method: 'post',
              input: z.object({
                regular: z.string(),
                min: z.string().min(1),
                max: z.string().max(15),
                range: z.string().min(2).max(3),
                email: z.string().email(),
                uuid: z.string().uuid(),
                cuid: z.string().cuid(),
                url: z.string().url(),
                numeric: z.string().regex(/\d+/),
                combined: z.string().nonempty().email().regex(/.*@example\.com/si).max(90)
              }),
              output: z.object({
                nonempty: z.string().nonempty()
              }),
              handler: jest.fn()
            })
          }
        },
        version: '3.4.5',
        title: 'Testing strings',
        serverUrl: 'http://example.com'
      }).getSpecAsYaml();
      expect(spec).toMatchSnapshot();
    });

    test('should handle tuples', () => {
      const spec = new OpenAPI({
        routing: {
          v1: {
            getSomething: defaultEndpointsFactory.build({
              method: 'post',
              input: z.object({
                ofOne: z.tuple([z.boolean()]),
                ofStrings: z.tuple([z.string(), z.string().nullable()]),
                complex: z.tuple([z.boolean(), z.string(), z.number().int().positive()])
              }),
              output: z.object({
                empty: z.tuple([])
              }),
              handler: jest.fn()
            })
          }
        },
        version: '3.4.5',
        title: 'Testing tuples',
        serverUrl: 'http://example.com'
      }).getSpecAsYaml();
      expect(spec).toMatchSnapshot();
    });

    test('should handle enum types', () => {
      const spec = new OpenAPI({
        routing: {
          v1: {
            getSomething: defaultEndpointsFactory.build({
              method: 'post',
              input: z.object({
                regularEnum: z.enum(['ABC', 'DEF']),
              }),
              output: z.object({
                nativeEnum: z.nativeEnum({FEG: 1, XYZ: 2})
              }),
              handler: async () => ({
                nativeEnum: 1
              })
            })
          }
        },
        version: '3.4.5',
        title: 'Testing enums',
        serverUrl: 'http://example.com'
      }).getSpecAsYaml();
      expect(spec).toMatchSnapshot();
    });

    test('should handle z.preprocess()', () => {
      const string = z.preprocess((arg) => String(arg), z.string());
      const number = z.preprocess((arg) => parseInt(String(arg), 16), z.number().int().nonnegative());
      const boolean = z.preprocess((arg) => !!arg, z.boolean());
      const spec = new OpenAPI({
        routing: {
          v1: {
            getSomething: defaultEndpointsFactory.build({
              method: 'get',
              input: z.object({ string, number }),
              output: z.object({ boolean }),
              handler: async () => ({
                boolean: [] as unknown as boolean // @todo check this out without type forcing in future Zod versions
              })
            })
          }
        },
        version: '3.4.5',
        title: 'Testing z.preprocess()',
        serverUrl: 'http://example.com'
      }).getSpecAsYaml();
      expect(spec).toMatchSnapshot();
      expect(string.parse(123)).toBe('123');
      expect(number.parse('0xFF')).toBe(255);
      expect(boolean.parse([])).toBe(true);
      expect(boolean.parse('')).toBe(false);
      expect(boolean.parse(null)).toBe(false);
    });

    test('should throw on unsupported types', () => {
      [
        z.undefined(),
        z.map(z.any(), z.any()),
        z.function(),
        z.lazy(() => z.any()),
        z.promise(z.any()),
        z.unknown(),
        z.never(),
        z.void()
      ].forEach((zodType) => {
        expect(() => new OpenAPI({
          routing: {
            v1: {
              getSomething: defaultEndpointsFactory.build({
                method: 'post',
                input: z.object({
                  property: zodType
                }),
                output: z.object({}),
                handler: async () => ({})
              })
            }
          },
          version: '3.4.5',
          title: 'Testing unsupported types',
          serverUrl: 'http://example.com'
        })).toThrowError(/Zod type Zod\w+ is unsupported/);
      });
    });
  });

  describe('Issue #98', () => {
    test('Should describe non-empty array', () => {
      // There is no such class as ZodNonEmptyArray in Zod v3.7.0+
      // It existed though in Zod v3.6.x:
      // @see https://github.com/colinhacks/zod/blob/v3.6.1/src/types.ts#L1204
      const spec = new OpenAPI({
        routing: {
          v1: {
            getSomething: defaultEndpointsFactory.build({
              methods: ['get', 'post'],
              input: z.object({
                arr: z.array(z.string()).nonempty(),
              }),
              output: z.object({
                arr: z.array(z.string()).nonempty()
              }),
              handler: async ({input}) => ({
                arr: input.arr
              })
            })
          }
        },
        version: '3.4.5',
        title: 'Testing issue #98',
        serverUrl: 'http://example.com'
      }).getSpecAsYaml();
      expect(spec).toMatchSnapshot();
    });

    test('should union schemas', () => {
      const baseSchema = z.object({ id: z.string() });
      const subType1 = baseSchema.extend({ field1: z.string() });
      const subType2 = baseSchema.extend({ field2: z.string() });
      const unionSchema = z.union([subType1, subType2]);
      type TestingType = z.infer<typeof unionSchema>;

      expectType<TestingType>({id: 'string', field1: 'string'});
      expectType<TestingType>({id: 'string', field2: 'string'});

      const spec = new OpenAPI({
        routing: {
          v1: {
            getSomething: defaultEndpointsFactory.build({
              method: 'post',
              input: unionSchema,
              output: unionSchema,
              handler: async ({input}) => {
                if ('field1' in input) {
                  return {
                    id: `test, ${input.id}`,
                    field1: input.field1
                  };
                }
                return {
                  id: 'other test',
                  field2: input.field2
                };
              }
            })
          }
        },
        version: '3.4.5',
        title: 'Testing issue #98',
        serverUrl: 'http://example.com'
      }).getSpecAsYaml();
      expect(spec).toMatchSnapshot();
    });
  });
});
