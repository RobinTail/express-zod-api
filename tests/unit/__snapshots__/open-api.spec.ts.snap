// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Open API generator generateOpenApi() should generate the correct schema for complex types 1`] = `
"openapi: 3.0.0
info:
  title: Testing Complex Types
  version: 3.4.5
paths:
  /v1/getSomething:
    get:
      responses:
        \\"200\\":
          description: GET /v1/getSomething Successful response
          content:
            application/json:
              schema:
                type: object
                properties:
                  status:
                    type: string
                    enum:
                      - success
                  data:
                    type: object
                    properties:
                      literal:
                        type: string
                        enum:
                          - something
                      transformation:
                        type: number
                    required:
                      - literal
                      - transformation
                required:
                  - status
                  - data
        \\"400\\":
          description: GET /v1/getSomething Error response
          content:
            application/json:
              schema:
                type: object
                properties:
                  status:
                    type: string
                    enum:
                      - error
                  error:
                    type: object
                    properties:
                      message:
                        type: string
                    required:
                      - message
                required:
                  - status
                  - error
      parameters:
        - name: array
          in: query
          required: true
          schema:
            type: array
            items:
              type: number
            description: GET /v1/getSomething parameter
        - name: transformer
          in: query
          required: true
          schema:
            type: string
            description: GET /v1/getSomething parameter
components:
  schemas: {}
  responses: {}
  parameters: {}
  examples: {}
  requestBodies: {}
  headers: {}
  securitySchemes: {}
  links: {}
  callbacks: {}
tags: []
servers:
  - url: http://example.com
"
`;

exports[`Open API generator generateOpenApi() should generate the correct schema for intersection type 1`] = `
"openapi: 3.0.0
info:
  title: Testing Intersection and And types
  version: 3.4.5
paths:
  /v1/getSomething:
    post:
      responses:
        \\"200\\":
          description: POST /v1/getSomething Successful response
          content:
            application/json:
              schema:
                type: object
                properties:
                  status:
                    type: string
                    enum:
                      - success
                  data:
                    type: object
                    properties:
                      and:
                        allOf:
                          - type: object
                            properties:
                              five:
                                type: number
                            required:
                              - five
                          - type: object
                            properties:
                              six:
                                type: string
                            required:
                              - six
                    required:
                      - and
                required:
                  - status
                  - data
        \\"400\\":
          description: POST /v1/getSomething Error response
          content:
            application/json:
              schema:
                type: object
                properties:
                  status:
                    type: string
                    enum:
                      - error
                  error:
                    type: object
                    properties:
                      message:
                        type: string
                    required:
                      - message
                required:
                  - status
                  - error
      requestBody:
        content:
          application/json:
            schema:
              type: object
              properties:
                intersection:
                  allOf:
                    - type: object
                      properties:
                        one:
                          type: string
                      required:
                        - one
                    - type: object
                      properties:
                        two:
                          type: string
                      required:
                        - two
              required:
                - intersection
              description: POST /v1/getSomething request body
components:
  schemas: {}
  responses: {}
  parameters: {}
  examples: {}
  requestBodies: {}
  headers: {}
  securitySchemes: {}
  links: {}
  callbacks: {}
tags: []
servers:
  - url: http://example.com
"
`;

exports[`Open API generator generateOpenApi() should generate the correct schema for nullable and optional types 1`] = `
"openapi: 3.0.0
info:
  title: Testing Nullable and Optional Types
  version: 3.4.5
paths:
  /v1/getSomething:
    get:
      responses:
        \\"200\\":
          description: GET /v1/getSomething Successful response
          content:
            application/json:
              schema:
                type: object
                properties:
                  status:
                    type: string
                    enum:
                      - success
                  data:
                    type: object
                    properties:
                      nullable:
                        nullable: true
                        type: string
                    required:
                      - nullable
                required:
                  - status
                  - data
        \\"400\\":
          description: GET /v1/getSomething Error response
          content:
            application/json:
              schema:
                type: object
                properties:
                  status:
                    type: string
                    enum:
                      - error
                  error:
                    type: object
                    properties:
                      message:
                        type: string
                    required:
                      - message
                required:
                  - status
                  - error
      parameters:
        - name: optional
          in: query
          required: false
          schema:
            type: string
            description: GET /v1/getSomething parameter
components:
  schemas: {}
  responses: {}
  parameters: {}
  examples: {}
  requestBodies: {}
  headers: {}
  securitySchemes: {}
  links: {}
  callbacks: {}
tags: []
servers:
  - url: http://example.com
"
`;

exports[`Open API generator generateOpenApi() should generate the correct schema for union type 1`] = `
"openapi: 3.0.0
info:
  title: Testing Union and Or Types
  version: 3.4.5
paths:
  /v1/getSomething:
    post:
      responses:
        \\"200\\":
          description: POST /v1/getSomething Successful response
          content:
            application/json:
              schema:
                type: object
                properties:
                  status:
                    type: string
                    enum:
                      - success
                  data:
                    type: object
                    properties:
                      or:
                        oneOf:
                          - type: string
                          - type: number
                    required:
                      - or
                required:
                  - status
                  - data
        \\"400\\":
          description: POST /v1/getSomething Error response
          content:
            application/json:
              schema:
                type: object
                properties:
                  status:
                    type: string
                    enum:
                      - error
                  error:
                    type: object
                    properties:
                      message:
                        type: string
                    required:
                      - message
                required:
                  - status
                  - error
      requestBody:
        content:
          application/json:
            schema:
              type: object
              properties:
                union:
                  oneOf:
                    - type: object
                      properties:
                        one:
                          type: string
                        two:
                          type: number
                      required:
                        - one
                        - two
                    - type: object
                      properties:
                        two:
                          type: number
                        three:
                          type: string
                      required:
                        - two
                        - three
              required:
                - union
              description: POST /v1/getSomething request body
components:
  schemas: {}
  responses: {}
  parameters: {}
  examples: {}
  requestBodies: {}
  headers: {}
  securitySchemes: {}
  links: {}
  callbacks: {}
tags: []
servers:
  - url: http://example.com
"
`;

exports[`Open API generator generateOpenApi() should generate the correct schema of example routing 1`] = `
"openapi: 3.0.0
info:
  title: Example API
  version: 1.2.3
paths:
  /v1/getUser:
    get:
      responses:
        \\"200\\":
          description: GET /v1/getUser Successful response
          content:
            application/json:
              schema:
                type: object
                properties:
                  status:
                    type: string
                    enum:
                      - success
                  data:
                    type: object
                    properties:
                      id:
                        type: number
                      name:
                        type: string
                    required:
                      - id
                      - name
                required:
                  - status
                  - data
        \\"400\\":
          description: GET /v1/getUser Error response
          content:
            application/json:
              schema:
                type: object
                properties:
                  status:
                    type: string
                    enum:
                      - error
                  error:
                    type: object
                    properties:
                      message:
                        type: string
                    required:
                      - message
                required:
                  - status
                  - error
      description: example user retrieval endpoint
      parameters:
        - name: id
          in: query
          required: true
          schema:
            type: string
            description: GET /v1/getUser parameter
  /v1/setUser:
    post:
      responses:
        \\"200\\":
          description: POST /v1/setUser Successful response
          content:
            application/json:
              schema:
                type: object
                properties:
                  status:
                    type: string
                    enum:
                      - success
                  data:
                    type: object
                    properties:
                      name:
                        type: string
                      timestamp:
                        type: number
                    required:
                      - name
                      - timestamp
                required:
                  - status
                  - data
        \\"400\\":
          description: POST /v1/setUser Error response
          content:
            application/json:
              schema:
                type: object
                properties:
                  status:
                    type: string
                    enum:
                      - error
                  error:
                    type: object
                    properties:
                      message:
                        type: string
                    required:
                      - message
                required:
                  - status
                  - error
      description: example user update endpoint
      requestBody:
        content:
          application/json:
            schema:
              type: object
              properties:
                key:
                  type: string
                id:
                  type: number
                name:
                  type: string
              required:
                - key
                - id
                - name
              description: POST /v1/setUser request body
  /v1/avatar:
    get:
      responses:
        \\"200\\":
          description: GET /v1/avatar Successful response
          content:
            image/svg+xml:
              schema:
                type: string
        \\"400\\":
          description: GET /v1/avatar Error response
          content:
            text/plain:
              schema:
                type: string
      parameters:
        - name: userId
          in: query
          required: true
          schema:
            type: string
            description: GET /v1/avatar parameter
  /v1/stream:
    get:
      responses:
        \\"200\\":
          description: GET /v1/stream Successful response
          content:
            image/*:
              schema:
                type: string
                format: binary
        \\"400\\":
          description: GET /v1/stream Error response
          content:
            text/plain:
              schema:
                type: string
      parameters:
        - name: userId
          in: query
          required: true
          schema:
            type: string
            description: GET /v1/stream parameter
components:
  schemas: {}
  responses: {}
  parameters: {}
  examples: {}
  requestBodies: {}
  headers: {}
  securitySchemes: {}
  links: {}
  callbacks: {}
tags: []
servers:
  - url: http://example.com
"
`;

exports[`Open API generator generateOpenApi() should handle bigint, boolean, date and null 1`] = `
"openapi: 3.0.0
info:
  title: Testing additional types
  version: 3.4.5
paths:
  /v1/getSomething:
    post:
      responses:
        \\"200\\":
          description: POST /v1/getSomething Successful response
          content:
            application/json:
              schema:
                type: object
                properties:
                  status:
                    type: string
                    enum:
                      - success
                  data:
                    type: object
                    properties:
                      \\"null\\":
                        nullable: true
                        type: string
                        format: \\"null\\"
                    required:
                      - \\"null\\"
                required:
                  - status
                  - data
        \\"400\\":
          description: POST /v1/getSomething Error response
          content:
            application/json:
              schema:
                type: object
                properties:
                  status:
                    type: string
                    enum:
                      - error
                  error:
                    type: object
                    properties:
                      message:
                        type: string
                    required:
                      - message
                required:
                  - status
                  - error
      requestBody:
        content:
          application/json:
            schema:
              type: object
              properties:
                bigint:
                  type: integer
                  format: int64
                boolean:
                  type: boolean
                date:
                  type: string
                  format: date
              required:
                - bigint
                - boolean
                - date
              description: POST /v1/getSomething request body
components:
  schemas: {}
  responses: {}
  parameters: {}
  examples: {}
  requestBodies: {}
  headers: {}
  securitySchemes: {}
  links: {}
  callbacks: {}
tags: []
servers:
  - url: http://example.com
"
`;

exports[`Open API generator generateOpenApi() should handle enum types 1`] = `
"openapi: 3.0.0
info:
  title: Testing enums
  version: 3.4.5
paths:
  /v1/getSomething:
    post:
      responses:
        \\"200\\":
          description: POST /v1/getSomething Successful response
          content:
            application/json:
              schema:
                type: object
                properties:
                  status:
                    type: string
                    enum:
                      - success
                  data:
                    type: object
                    properties:
                      nativeEnum:
                        type: number
                        enum:
                          - 1
                          - 2
                    required:
                      - nativeEnum
                required:
                  - status
                  - data
        \\"400\\":
          description: POST /v1/getSomething Error response
          content:
            application/json:
              schema:
                type: object
                properties:
                  status:
                    type: string
                    enum:
                      - error
                  error:
                    type: object
                    properties:
                      message:
                        type: string
                    required:
                      - message
                required:
                  - status
                  - error
      requestBody:
        content:
          application/json:
            schema:
              type: object
              properties:
                regularEnum:
                  type: string
                  enum:
                    - ABC
                    - DEF
              required:
                - regularEnum
              description: POST /v1/getSomething request body
components:
  schemas: {}
  responses: {}
  parameters: {}
  examples: {}
  requestBodies: {}
  headers: {}
  securitySchemes: {}
  links: {}
  callbacks: {}
tags: []
servers:
  - url: http://example.com
"
`;

exports[`Open API generator generateOpenApi() should handle transformation schema in output 1`] = `
"openapi: 3.0.0
info:
  title: Testing Transformation in response schema
  version: 3.4.5
paths:
  /v1/getSomething:
    post:
      responses:
        \\"200\\":
          description: POST /v1/getSomething Successful response
          content:
            application/json:
              schema:
                type: object
                properties:
                  status:
                    type: string
                    enum:
                      - success
                  data:
                    type: object
                    properties:
                      transform:
                        type: number
                    required:
                      - transform
                required:
                  - status
                  - data
        \\"400\\":
          description: POST /v1/getSomething Error response
          content:
            application/json:
              schema:
                type: object
                properties:
                  status:
                    type: string
                    enum:
                      - error
                  error:
                    type: object
                    properties:
                      message:
                        type: string
                    required:
                      - message
                required:
                  - status
                  - error
      requestBody:
        content:
          application/json:
            schema:
              type: object
              properties:
                one:
                  type: string
                two:
                  type: number
              required:
                - one
                - two
              description: POST /v1/getSomething request body
components:
  schemas: {}
  responses: {}
  parameters: {}
  examples: {}
  requestBodies: {}
  headers: {}
  securitySchemes: {}
  links: {}
  callbacks: {}
tags: []
servers:
  - url: http://example.com
"
`;
